-1

Tested at: 96d5d1b104b8c75f1fc227c3e748af6b4294f15f

CMSSW: CMSSW_11_2_X_2020-07-01-2300
SCRAM_ARCH: slc7_amd64_gcc820

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
 - #30236 @mandrenguyen: Modified photon selection for heavy-ion miniAOD
 - #30423 @bsunanda: Phase2-hgx250 Comparison of results from flat file and from insitu checks
 - #30496 @cms-sw: Added missing include to fix the ROOT IBs build errors
 - #30459 @JamminJones: added esConsumes to modules in EventFilter/RPCRawToDigi
 - #30344 @dtp2-tpg-am: DT TPG for Phase-2
 - #30127 @aminnj: TBLayer computeCrossings doesn't allow for some displaced starting points
 - #30446 @duartej: [Pixel 3D Digitizer] Fix wrong channel assignation
 - #30476 @camolezi: Reduce boost dependency in DQM/TrackingMonitor 
 - #30475 @JamminJones: added esConsumes to modules in EventFilter/SiStripRawToDigi
 - #30474 @JamminJones: added esConsumes to modules on EventFilter/SiPixelRawToDigi

You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-49e10f/7608/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-49e10f/7608/git-merge-result

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-49e10f/7608/summary.html

I found follow errors while testing this PR

**Failed tests:  Build  ClangBuild**

* **Build**:

I found compilation error when building: 

<pre>In file included from /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_2_X_2020-07-01-2300/src/DataFormats/PatCandidates/interface/Lepton.h:22,
                 from /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_2_X_2020-07-01-2300/src/DataFormats/PatCandidates/interface/Muon.h:28,
                 from /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_2_X_2020-07-01-2300/src/PhysicsTools/PatAlgos/test/private/PATUserDataTestModule.cc:33:
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_2_X_2020-07-01-2300/src/DataFormats/PatCandidates/interface/PATObject.h: In instantiation of 'const string& pat::PATObject<ObjectType>::userDataObjectType(const string&) const [with ObjectType = reco::Muon; std::__cxx11::string = std::__cxx11::basic_string<char>]':
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_2_X_2020-07-01-2300/src/PhysicsTools/PatAlgos/test/private/PATUserDataTestModule.cc:144:77:   required from here
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_2_X_2020-07-01-2300/src/DataFormats/PatCandidates/interface/PATObject.h:332:54: error: returning reference to temporary [-Werror=return-local-addr]
       return (data != nullptr ? data->typeName() : "");
                                                      ^
cc1plus: some warnings being treated as errors
gmake: *** [tmp/slc7_amd64_gcc820/src/PhysicsTools/PatAlgos/test/PhysicsToolsPatAlgos_testAnalyzers/private/PATUserDataTestModule.cc.o] Error 1
>> Building  edm plugin tmp/slc7_amd64_gcc820/src/PhysicsTools/PatAlgos/test/PhysicsToolsPatAlgos_testAnalyzers/libPhysicsToolsPatAlgos_testAnalyzers.so
</pre>

* **Clang**:

I found compilation warning while trying to compile with clang. Command used:
```
USER_CUDA_FLAGS='--expt-relaxed-constexpr' USER_CXXFLAGS='-Wno-register -fsyntax-only' scram build -k -j 32 COMPILER='llvm compile'

```
 See details on the summary page.


The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
 - #30236 @mandrenguyen: Modified photon selection for heavy-ion miniAOD
 - #30423 @bsunanda: Phase2-hgx250 Comparison of results from flat file and from insitu checks
 - #30496 @cms-sw: Added missing include to fix the ROOT IBs build errors
 - #30459 @JamminJones: added esConsumes to modules in EventFilter/RPCRawToDigi
 - #30344 @dtp2-tpg-am: DT TPG for Phase-2
 - #30127 @aminnj: TBLayer computeCrossings doesn't allow for some displaced starting points
 - #30446 @duartej: [Pixel 3D Digitizer] Fix wrong channel assignation
 - #30476 @camolezi: Reduce boost dependency in DQM/TrackingMonitor 
 - #30475 @JamminJones: added esConsumes to modules in EventFilter/SiStripRawToDigi
 - #30474 @JamminJones: added esConsumes to modules on EventFilter/SiPixelRawToDigi

You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-49e10f/7608/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-49e10f/7608/git-merge-result

--- a/DataFormats/PatCandidates/interface/PATObject.h
+++ b/DataFormats/PatCandidates/interface/PATObject.h
@@ -38,11 +38,8 @@
 #include "FWCore/MessageLogger/interface/MessageLogger.h"
 
 namespace pat {
-
-  namespace pat_statics {
-    static const reco::CandidatePtrVector EMPTY_CPV;
-    static const std::string EMPTY_STR("");
-  }  // namespace pat_statics
+  const reco::CandidatePtrVector &get_empty_cpv();
+  const std::string &get_empty_str();
 
   template <class ObjectType>
   class PATObject : public ObjectType {
@@ -333,7 +330,7 @@ namespace pat {
     /// Get human-readable type of user data object, for debugging
     const std::string &userDataObjectType(const std::string &key) const {
       const pat::UserData *data = userDataObject_(key);
-      return (data != nullptr ? data->typeName() : pat_statics::EMPTY_STR);
+      return (data != nullptr ? data->typeName() : get_empty_str());
     };
     /// Get list of user data object names
     const std::vector<std::string> &userDataNames() const { return userDataLabels_; }
@@ -829,7 +826,7 @@ namespace pat {
   const reco::CandidatePtrVector &PATObject<ObjectType>::overlaps(const std::string &label) const {
     auto match = std::lower_bound(overlapLabels_.cbegin(), overlapLabels_.cend(), label);
     if (match == overlapLabels_.cend() || *match != label)
-      return pat_statics::EMPTY_CPV;
+      return get_empty_cpv();
     return overlapItems_[std::distance(overlapLabels_.begin(), match)];
   }
 

--- a/DataFormats/PatCandidates/src/PATObject.cc
+++ b/DataFormats/PatCandidates/src/PATObject.cc
@@ -1,0 +1,13 @@
+#include "DataFormats/PatCandidates/interface/PATObject.h"
+
+namespace pat {
+  const reco::CandidatePtrVector& get_empty_cpv() {
+    static const reco::CandidatePtrVector EMPTY_CPV;
+    return EMPTY_CPV;
+  }
+
+  const std::string& get_empty_str() {
+    static const std::string EMPTY_STR;
+    return EMPTY_STR;
+  }
+}  // namespace pat

